File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        
        ans[0] = 1;
        int pre = 1;
        // pre
        for (int i = 1; i < nums.length; i++) {
            pre = pre * nums[i - 1];
            ans[i] = pre;
        }
        // post
        int post = 1;
        for (int i = nums.length - 2; i >= 0; i--) {
            post = post * nums[i + 1];
            ans[i] = post * ans[i];
        }
        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pre, post = [1] * n, [1] * n
        
        for i in range(1, n):
            pre[i] = nums[i - 1] * pre[i - 1]
        
        for i in range(n - 2, -1, -1):
            post[i] = nums[i + 1] * post[i + 1]
        
        res = []
        for i in range(n):
            res.append(pre[i] * post[i])
        
        return res